src/satosa/micro_services/idp_metadata_attribute_store.py

```diff
@@ -0,0 +1,229 @@
+            idp_config = copy.deepcopy(IdpMetadataAttributeStore.config_defaults)
+            if 'default' in self.config:
+                idp_config.update(self.config['default'])
+            idp_config.update(config[idp])
+
+            self.config[idp] = idp_config
+
+        satosa_logging(logger, logging.INFO, "IdP Metadata Attribute Store microservice initialized", None)
+
+    def _first_lang_element_text(self, elements, lang='en'):
+        """
+        Loop over the list representing XML elements that contain text and find
+        the first text value for the input lang where 'en' or English is the
+        default lang.
+
+        Each item in the list is a dictionary with keys
+
+            __class__
+            lang
+            text
+
+        as expected from the metadata returned for an entity by the MetadataStore
+        class from pysaml2.
+
+        If no element has the input lang then return the text from the first 
+        element.
+
+        If no element has text then return an empty string.
+        """
+        for e in elements:
+            if lang in e:
+                if 'text' in e:
+                    return e['text']
+
+        for e in elements:
+            if 'text' in e:
+                return e['text']
+
+        return ''
+
+    def process(self, context, data):
+        """
+        Default interface for microservices. Process the input data for
+        the input context.
+        """
+        self.context = context
+
+        # Find the entityID for the IdP that issued the assertion.
+        try:
+            idp_entity_id = data.to_dict()['auth_info']['issuer']
+        except KeyError as err:
+            satosa_logging(logger, logging.ERROR, "Unable to determine the entityID for the IdP issuer", context.state)
+            return super().process(context, data)
+
+        # Get the configuration for the IdP.
+        if idp_entity_id in self.config.keys():
+            config = self.config[idp_entity_id]
+        else:
+            config = self.config['default']
+
+        satosa_logging(logger, logging.DEBUG, "Using config {}".format(config), context.state)
+
+        # Ignore this IdP if so configured.
+        if config['ignore']:
+            satosa_logging(logger, logging.INFO, "Ignoring IdP {}".format(idp_entity_id), context.state)
+            return super().process(context, data)
+
+        # Get the metadata store the SP for the proxy is using. This
+        # will be an instance of the class MetadataStore from mdstore.py
+        # in pysaml2.
+        metadata_store = context.internal_data['metadata_store']
+
+        # Get the metadata for the IdP.
+        try:
+            metadata = metadata_store[idp_entity_id]
+        except Exception as err:
+            satosa_logging(logger, logging.ERROR, "Unable to retrieve metadata for IdP {}".format(idp_entity_id), context.state)
+            return super().process(context, data)
```

Comment:
This is a recurring pattern:

```py
try:
  something()
except:
  return super().process(context, data)
```

This can be handled better by changing `process()` to call `_process()` as such:
```py
class IdpMetadataAttributeStoreWarning(SATOSAError):
  pass

def process(..):
  try:
    self._process(..)
  except IdpMetadataAttributeStoreWarning as ex:
    satosa_logging(logger, logging.WARN, str(ex), context.state)
    return super().process(context, data)

def _process(..):
  # all the code previously on process() but with modified parts as such:
  try:
    something()
  except SomeError as err:
    msg_warn = "something went wrong, but it's ok"
    raise IdpMetadataAttributeStoreWarning(msg_warn) from err
```

Errors of type `IdpMetadataAttributeStoreWarning` will be logged and let the process continue, but other errors will be raised and stop the execution. This also makes errors that we handle explicit.